src/NuGet.Core/NuGet.Versioning/PublicAPI.Shipped.txt

```diff
@@ -111,7 +111,7 @@ NuGet.Versioning.VersionRangeBase.Satisfies(NuGet.Versioning.NuGetVersion! versi
 NuGet.Versioning.VersionRangeBase.VersionRangeBase(NuGet.Versioning.NuGetVersion? minVersion = null, bool includeMinVersion = true, NuGet.Versioning.NuGetVersion? maxVersion = null, bool includeMaxVersion = false) -> void
 NuGet.Versioning.VersionRangeComparer
 NuGet.Versioning.VersionRangeComparer.Equals(NuGet.Versioning.VersionRangeBase? x, NuGet.Versioning.VersionRangeBase? y) -> bool
-NuGet.Versioning.VersionRangeComparer.GetHashCode(NuGet.Versioning.VersionRangeBase! obj) -> int
+NuGet.Versioning.VersionRangeComparer.GetHashCode(NuGet.Versioning.VersionRangeBase? obj) -> int
```

Comment:
I still think we should prefer API design over implementation and not make this nullable. 0 is a valid hash code for a non-null instance, and I think we should encourage API customers to not request nonsensical values. When I checkout this branch locally and build there are no build errors.

Reply:
> I still think we should prefer API design over implementation and not make this nullable. 0 is a valid hash code for a non-null instance.

I agree, but the current implementation already returns `0` when null value is passed. I think the changes proposed in the PR makes it explicit that this value can be null. Please let me know if I am missing some context here.

Reply:
My first sentence, that you quoted, states my position:

> I still think we should prefer API design over implementation and not make this nullable.

We already have a bunch of bad APIs (mostly too many public APIs) which makes some bug fixes and perf improvements difficult or impossible. We've had insertions into the .NET SDK fail in the past when we broke an API we didn't think anyone outside of NuGet would use, but the .NET SDK itself did. Let's not make things worse by annotating a badly designed API as allowing a null when it should have been designed & implemented not to.

Honestly, I think a `GetHashCode(T object)` itself is a bad design. If someone wanted GetHashCode for version without build metadata + GetHashCode without build metadata or prerelease, plus the "everything" GetHashCode, then `NuGetVersion.GetHashCode(HashCodeType)` (or even `GetHashCode(VersionComparer)`) would have been better, and would have been a similar API design to the existing `string.GetHashCode(StringComparer)`. But that's not what shipped 6+ years ago, and who knows how many people we'd break if we changed it now.

Anyway, we only added nullable annotations in 6.7, so they're brand new, not even out in GA yet. I wanted to take the opportunity to add some "API design" into the nullable annotations. I also understand wanting to annotate was was implemented, so don't take API design into consideration for existing APIs we annotate. Overall it's up to the team, I'm just one person. The only reason I didn't approve the PR is because a bunch of comparer allocations were missed in `PackageIdentityComparer.cs`.

Reply:
I don't feel strongly either way, so let me know what the team decides and we can update. Feel free to push updates here too if you prefer.

I understand the desire to try and tighten up the API design via the annotations.

Equally if a consumer today is passing null, then when they update they'll start getting a warning for code that arguably "works",  which is inconvenient.

Reply:
Thanks for the detailed explanation, @zivkan. I agree with reverting the `GetHashCode` related changes from this PR.